### PR TITLE
support for INSERT query

### PR DESCRIPTION
**Pull request by simakmi**

### CODE REVIEW

File: src/main/java/com/axibase/tsd/driver/jdbc/ext/AtsdPreparedStatement.java [hunk 1]
```diff
@@ -330,4 +336,97 @@ public void setTimeExpression(int parameterIndex, String value) throws SQLExcept
...
+		return copy;
+	}
+
+	@Override
+	public Meta.StatementType getStatementType() {
+		return getSignature() == null ? null : getSignature().statementType;
+	}
+
+	@Override
+	public ResultSetMetaData getMetaData() {
```
Comment (raipc) [inline src/main/java/com/axibase/tsd/driver/jdbc/ext/AtsdPreparedStatement.java]: I think it should be in AtsdMeta#prepare method

File: src/main/java/com/axibase/tsd/driver/jdbc/ext/AtsdPreparedStatement.java [hunk 2]
```diff
@@ -330,4 +336,97 @@ public void setTimeExpression(int parameterIndex, String value) throws SQLExcept
...
+						return result;
+					} catch (SQLException exc) {
+						throw new AtsdRuntimeException("", exc);
+					}
+				}
+				default: throw new UnsupportedOperationException("Not yet implemented for statement type: " + getStatementType());
+			}
+		}
+
+		return super.getMetaData();
```
Comment (raipc) [inline src/main/java/com/axibase/tsd/driver/jdbc/ext/AtsdPreparedStatement.java]: NPE in all resultSet.getSmth methods. Probably, we should implicitly throw an exception if column metadata cannot be resolved from SQL. (in AtsdMeta#prepare())

File: pom.xml [hunk 3]
```diff
@@ -554,12 +608,12 @@
 			</build>
 		</profile>
 	</profiles>
-	<distributionManagement>
-		<repository>
-			<id>sonatype</id>
-			<url>https://oss.sonatype.org/service/local/staging/deploy/maven2</url>
-		</repository>
-	</distributionManagement>
+	<!--<distributionManagement>-->
```
Comment (raipc) [inline pom.xml]: It is needed for publishing the artifacts to Maven Central

Comment (simakmi) [inline pom.xml]: I forgot to uncomment it. I fixed it.

File: src/main/java/com/axibase/tsd/driver/jdbc/ext/AtsdPreparedStatement.java [hunk 4]
```diff
@@ -364,12 +364,16 @@ public void addBatch(String sql) throws SQLException {
 	@Override
 	public ResultSetMetaData getMetaData() {
 		logger.debug("[getMetaData] executed: {} signature: {}", executed, getSignature());
-		if (executed) {
-			return super.getMetaData();
-		} else if (getSignature() != null && getSignature().columns == null) {
+
+		if (!executed && getSignature() != null && getSignature().columns == null) {
 			switch (getStatementType()) {
```
Comment (simakmi) [inline src/main/java/com/axibase/tsd/driver/jdbc/ext/AtsdPreparedStatement.java]: I agree. I'll replace it with simple if-else, because now we use only SELECT statement.
In this case if 'getStatementType' method returns null - it means that signature object is invalid. I'll add checking.

File: src/main/java/com/axibase/tsd/driver/jdbc/ext/AtsdPreparedStatement.java (hunk 1 above)
Comment (raipc) [inline src/main/java/com/axibase/tsd/driver/jdbc/ext/AtsdPreparedStatement.java]: Update: query can be complex and hard to parse. Let's for now drop support for this feature and throw SQLFeatureNotSupportedException, otherwise some instruments get confused.

Comment (simakmi) [inline src/main/java/com/axibase/tsd/driver/jdbc/ext/AtsdPreparedStatement.java]: I agree, but If we'll remove it - 'INSERT' statement will not work in MathLab, because it calls 'getMetaData' method and doesn't call 'execute' method before.
I added this feature for MathLab, because it generates simple query.

Comment (raipc) [inline src/main/java/com/axibase/tsd/driver/jdbc/ext/AtsdPreparedStatement.java]: does it work if getMetadata() returns null when statement type is not SELECT? Other databases like DB2 do this way.

Comment (simakmi) [inline src/main/java/com/axibase/tsd/driver/jdbc/ext/AtsdPreparedStatement.java]: We parse only SELECT statement, for other statement type we can return null.
MathLab executes metadata.getColumnCount() and other 'metadata.getColumn*' methods to determine how to generate INSERT query.

Comment (simakmi) [inline src/main/java/com/axibase/tsd/driver/jdbc/ext/AtsdPreparedStatement.java]: Later we can add SQL parser for any queries to generate MetaData.

File: src/main/java/com/axibase/tsd/driver/jdbc/ext/AtsdPreparedStatement.java (hunk 2 above)
Comment (simakmi) [inline src/main/java/com/axibase/tsd/driver/jdbc/ext/AtsdPreparedStatement.java]: I'll add AtsdResultSetMetaData. It will throw IllegalStateException if there is no columns.

Comment (raipc) [inline src/main/java/com/axibase/tsd/driver/jdbc/ext/AtsdPreparedStatement.java]: Maybe fall back to resultSet.getMetadata() if statement is executed, otherwise throw SQLFeatureNotSupportedException?

Comment (simakmi) [inline src/main/java/com/axibase/tsd/driver/jdbc/ext/AtsdPreparedStatement.java]: Yes, we can use resultSet,getMetadata() if statement is executed, otherwise we can throw SQLFeatureNotSupportedException if signature.columns == null.

Comment (simakmi) [inline src/main/java/com/axibase/tsd/driver/jdbc/ext/AtsdPreparedStatement.java]: 'getMetaData' doesn't allow to throw checked exception, we can use AtsdRuntimeException

Comment (raipc) [inline src/main/java/com/axibase/tsd/driver/jdbc/ext/AtsdPreparedStatement.java]: I have read the MATLAB code for data inserts. Yes, for now let's support such simple scenarios with select from single table

Comment (raipc) [inline src/main/java/com/axibase/tsd/driver/jdbc/ext/AtsdPreparedStatement.java]: Can lombok deal with it?

Comment (simakmi) [inline src/main/java/com/axibase/tsd/driver/jdbc/ext/AtsdPreparedStatement.java]: Lombok can, but we don't use lombok in this project. I think runtime exception is sufficient.

Comment (raipc) [inline src/main/java/com/axibase/tsd/driver/jdbc/ext/AtsdPreparedStatement.java]: I would better start using lombok. It's a compile-time dependency and will add nothing to jar size. 
The clients may catch only checked SQLExceptions, and unchecked exception may break the workflow since SQLException is declared in the interface.

File: src/main/java/com/axibase/tsd/driver/jdbc/ext/AtsdPreparedStatement.java (hunk 4 above)
Comment (simakmi) [inline src/main/java/com/axibase/tsd/driver/jdbc/ext/AtsdPreparedStatement.java]: done

File: src/main/java/com/axibase/tsd/driver/jdbc/ext/AtsdConnection.java [hunk 5]
```diff
@@ -69,14 +71,38 @@ protected ResultSet executeQueryInternal(AvaticaStatement statement, Meta.Signat
 			return super.executeQueryInternal(statement, signature, firstFrame, state, isUpdate);
 		} catch (SQLException e) {
 			throw ExceptionsUtil.unboxException(e);
+		} finally {
+			if (statement instanceof AtsdPreparedStatement) {
+				((AtsdPreparedStatement) statement).executed = true;
```
Comment (simakmi) [inline src/main/java/com/axibase/tsd/driver/jdbc/ext/AtsdConnection.java]: Query parsing moved to AtsdMeta#prepare method.

File: src/main/java/com/axibase/tsd/driver/jdbc/ext/AtsdMeta.java [hunk 6]
```diff
@@ -551,4 +646,106 @@ public void rollback(ConnectionHandle ch) {
...
+			} else {
+				result.add(value == null ? null : String.valueOf(value));
+			}
+		}
+
+		log.debug("[preparedValues] {}", result);
+		return result;
+	}
+
+	private static List<ColumnMetaData> prepareColumnMetaData(final StatementType statementType, final String query) {
```
Comment (raipc) [inline src/main/java/com/axibase/tsd/driver/jdbc/ext/AtsdMeta.java]: Queries may be complex and hard to parse. Probably we can do lazy init and query ATSD for metadata if it's needed. There is an undocumented feature: HEAD request will return metadata in header which can be parsed with MetadataRetriever#retrieveJsonSchemeFromHeader method.

Curl example:
curl --insecure  --verbose -u axibase:axibase --request HEAD  https://localhost:8443/api/sql?q=SELECT%20entity%20as%20%27atsd%20entity%27%2C%20t.value%2C%20datetime%20FROM%20jvm_memory_used%20t%20WHERE%20datetime%20%3E%20now%20-%201%2AMINUTE

Comment (simakmi) [inline src/main/java/com/axibase/tsd/driver/jdbc/ext/AtsdMeta.java]: I agree, because I did it for simple query, but now we need to prepare columns for complex queries. The undocumented feature is what I need.
